internal/projector/slide/user_test.go

```
@@ -107,9 +143,57 @@ func TestUser(t *testing.T) {
 				"user/1/first_name",
 				"user/1/last_name",
 				"user/1/structure_level_$",
-				"user/1/structure_level_$1",
+				"user/1/default_structure_level",
 			}
+			expectedKeys = append(expectedKeys, tt.addkeysexpected...)
 			assert.ElementsMatch(t, keys, expectedKeys)
 		})
 	}
 }
+
+func TestUserWithoutMeeting(t *testing.T) {
+	userSlide := setup(t)
+	closed := make(chan struct{})
+	defer close(closed)
+	data := map[string]string{
+		"user/1/username":                `"jonny123"`,
+		"user/1/title":                   `"Dr."`,
+		"user/1/first_name":              `"Jonny"`,
+		"user/1/last_name":               `"Bo"`,
+		"user/1/default_structure_level": `"Switzerland"`,
+	}
+
+	ds := dsmock.NewMockDatastore(closed, data)
+
+	p7on := &projector.Projection{
+		ContentObjectID: "user/1",
+	}
+
+	bs, keys, err := userSlide.Slide(context.Background(), ds, p7on)
+	assert.NoError(t, err)
+	assert.JSONEq(t, `{"user":"Dr. Jonny Bo (Switzerland)"}`, string(bs))
+	expectedKeys := []string{"user/1/username", "user/1/title", "user/1/first_name", "user/1/last_name", "user/1/default_structure_level", "user/1/structure_level_$"}
+	assert.ElementsMatch(t, keys, expectedKeys)
+}
+
+func TestUserWithDataWrapError(t *testing.T) {
+	userSlide := setup(t)
+	closed := make(chan struct{})
+	defer close(closed)
+	data := map[string]string{
+		"user/1/username": "jonny123",
+	}
+
+	ds := dsmock.NewMockDatastore(closed, data)
+
+	p7on := &projector.Projection{
+		ContentObjectID: "user/1",
+		MeetingID:       222,
+	}
+
+	bs, keys, err := userSlide.Slide(context.Background(), ds, p7on)
+	assert.Nil(t, bs)
+	assert.Nil(t, keys)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "Error encoding responceData `map[user:map[1:map[username:jonny123]]]`")
```

I don't understand why this should be an error. I thought this would return `{"user":"jonny123"}` Can you explain this to me.

---

The reason why it's raising the error are the missing  around the "jonny123"-data. I just wanted to produce an error for testing purposes.
Choose another more serious error: based on new specification in [wiki](https://github.com/OpenSlides/OpenSlides/wiki/Projector-System#user) if there isn't minimum one of username, firstName, lastName there is no output and we should raise an error.

---

I would not test the error string. I think the strings are for humans and the test should not fail if the string changes.

If you just want to test that an error happens, then you can check that error is not nil.

If you want to check for a specific error, you should check the error type. `error` is just an interface. Everything with a method `Error() string` can be an error. Most go developers create errors and export them: https://golang.org/src/net/http/server.go?s=64501:64553#L41

I don't line that the error-objects are varialbe (see [this blog post](https://dave.cheney.net/2016/04/07/constant-errors)). And I also don't like to export error objects. This creates unnecessary dependencies between packages. For this reason, I identify errors by methods. See https://github.com/OpenSlides/openslides-vote-service/blob/main/internal/backends/test/test.go#L43-L46

---

I'm using now an explicit interface for errors in slides-function. The error will be returned in the expected client format. okay, I expect Finn will agree. What do you think about this way?
The provoked error itself we can remove, because the username since some days is really required. 

About checking the error string I don't agree. We should always check it if there is no own type, constant etc. to identify. Otherwise the test may fail with a new error of the same class without being discovered.